File: shared/playslike/PlaysLikeService.kt
package com.golfiq.shared.playslike

import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.round
import org.json.JSONObject

object PlaysLikeService {
    data class ElevationProviderData(
        val elevationM: Double,
        val ttlSeconds: Int,
        val etag: String?,
    )

    data class WindProviderData(
        val speedMps: Double,
        val dirFromDeg: Double,
        val wParallel: Double?,
        val wPerp: Double?,
        val ttlSeconds: Int,
        val etag: String?,
    )

    private data class CacheEntry<T>(
        var value: T,
        var etag: String?,
        var expiresAt: Long,
    )

    @Volatile
    private var providersBaseUrl: String? = null

    private val elevationCache = ConcurrentHashMap<String, CacheEntry<ElevationProviderData>>()
    private val windCache = ConcurrentHashMap<String, CacheEntry<WindProviderData>>()
    private const val MPS_TO_MPH = 2.237
    private const val MPH_TO_MPS = 1.0 / 2.237

    fun mpsToMph(value: Double): Double = value * MPS_TO_MPH

    fun mphToMps(value: Double): Double = value * MPH_TO_MPS

    private fun sanitizeDistance(value: Double): Double =
        if (value.isFinite() && value > 0.0) value else 0.0

    private fun roundTo(value: Double, decimals: Int): Double {
        if (!value.isFinite()) return 0.0
        val factor = 10.0.pow(decimals.coerceAtLeast(0))
        return round(value * factor) / factor
    }

    fun setProvidersBaseUrl(url: String?) {
        providersBaseUrl = url?.trimEnd('/')
    }

    fun getProvidersBaseUrl(): String? = providersBaseUrl

    fun fetchElevation(lat: Double, lon: Double): ElevationProviderData {
        val base = providersBaseUrl ?: return ElevationProviderData(0.0, 0, null)
        val key = cacheKey(lat, lon)
        val cached = elevationCache[key]
        val now = nowMillis()
        if (cached != null && cached.expiresAt > now) {
            return cached.value.copy(etag = cached.etag)
        }

        val url = buildUrl(
            base,
            "/providers/elevation",
            mapOf("lat" to formatCoord(lat), "lon" to formatCoord(lon)),
        )
        val connection = openConnection(url, cached?.etag)
        try {
            val status = connection.responseCode
            if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
                if (cached == null) {
                    throw IllegalStateException("304 without cached elevation entry")
                }
                val etag = stripWeakEtag(connection.getHeaderField("ETag")) ?: cached.etag
                val ttl = parseMaxAge(connection.getHeaderField("Cache-Control")) ?: cached.value.ttlSeconds
                cached.value = cached.value.copy(ttlSeconds = ttl, etag = null)
                cached.etag = etag
                cached.expiresAt = now + ttl * 1000L
                elevationCache[key] = cached
                return cached.value.copy(etag = cached.etag)
            }

            if (status !in 200..299) {
                throw IllegalStateException("Elevation provider error: $status")
            }

            val body = readBody(connection)
            val json = JSONObject(body)
            val ttlCandidate = json.optInt("ttl_s", -1)
            val ttl = if (ttlCandidate >= 0) ttlCandidate else parseMaxAge(connection.getHeaderField("Cache-Control")) ?: 0
            val etag = json.optString("etag", null) ?: stripWeakEtag(connection.getHeaderField("ETag"))
            val data = ElevationProviderData(
                elevationM = json.optDouble("elevation_m", 0.0),
                ttlSeconds = ttl,
                etag = etag,
            )
            elevationCache[key] = CacheEntry(
                value = data.copy(etag = null),
                etag = etag,
                expiresAt = now + ttl * 1000L,
            )
            return data
        } finally {
            connection.disconnect()
        }
    }

    fun fetchWind(lat: Double, lon: Double, bearing: Double? = null): WindProviderData {
        val base = providersBaseUrl ?: return WindProviderData(0.0, 0.0, 0.0, 0.0, 0, null)
        val key = cacheKey(lat, lon) + (bearing?.let { String.format(Locale.US, "@%.2f", it) } ?: "")
        val cached = windCache[key]
        val now = nowMillis()
        if (cached != null && cached.expiresAt > now) {
            return cached.value.copy(etag = cached.etag)
        }

        val params = mutableMapOf(
            "lat" to formatCoord(lat),
            "lon" to formatCoord(lon),
        )
        if (bearing != null) {
            params["bearing"] = String.format(Locale.US, "%.2f", bearing)
        }
        val url = buildUrl(base, "/providers/wind", params)
        val connection = openConnection(url, cached?.etag)
        try {
            val status = connection.responseCode
            if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
                if (cached == null) {
                    throw IllegalStateException("304 without cached wind entry")
                }
                val etag = stripWeakEtag(connection.getHeaderField("ETag")) ?: cached.etag
                val ttl = parseMaxAge(connection.getHeaderField("Cache-Control")) ?: cached.value.ttlSeconds
                cached.value = cached.value.copy(ttlSeconds = ttl, etag = null)
                cached.etag = etag
                cached.expiresAt = now + ttl * 1000L
                windCache[key] = cached
                return cached.value.copy(etag = cached.etag)
            }

            if (status !in 200..299) {
                throw IllegalStateException("Wind provider error: $status")
            }

            val body = readBody(connection)
            val json = JSONObject(body)
            val ttlCandidate = json.optInt("ttl_s", -1)
            val ttl = if (ttlCandidate >= 0) ttlCandidate else parseMaxAge(connection.getHeaderField("Cache-Control")) ?: 0
            val etag = json.optString("etag", null) ?: stripWeakEtag(connection.getHeaderField("ETag"))
            val wParallel = if (json.isNull("w_parallel")) null else json.optDouble("w_parallel")
            val wPerp = if (json.isNull("w_perp")) null else json.optDouble("w_perp")
            val data = WindProviderData(
                speedMps = json.optDouble("speed_mps", 0.0),
                dirFromDeg = json.optDouble("dir_from_deg", 0.0),
                wParallel = wParallel,
                wPerp = wPerp,
                ttlSeconds = ttl,
                etag = etag,
            )
            windCache[key] = CacheEntry(
                value = data.copy(etag = null),
                etag = etag,
                expiresAt = now + ttl * 1000L,
            )
            return data
        } finally {
            connection.disconnect()
        }
    }

    private fun cacheKey(lat: Double, lon: Double): String =
        String.format(Locale.US, "%.5f,%.5f", lat, lon)

    private fun formatCoord(value: Double): String =
        String.format(Locale.US, "%.5f", value)

    private fun nowMillis(): Long = System.currentTimeMillis()

    private fun buildUrl(base: String, path: String, params: Map<String, String>): String {
        val query = params.entries.joinToString("&") { (key, value) ->
            "%s=%s".format(
                Locale.US,
                encode(key),
                encode(value),
            )
        }
        return "$base$path${if (query.isEmpty()) "" else "?$query"}"
    }

    private fun encode(value: String): String =
        java.net.URLEncoder.encode(value, StandardCharsets.UTF_8.name())

    private fun openConnection(url: String, etag: String?): HttpURLConnection {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = 10_000
        connection.readTimeout = 10_000
        if (!etag.isNullOrEmpty()) {
            connection.setRequestProperty("If-None-Match", etag)
        }
        return connection
    }

    private fun readBody(connection: HttpURLConnection): String {
        val stream = if (connection.responseCode >= 400) connection.errorStream else connection.inputStream
        BufferedReader(InputStreamReader(stream, StandardCharsets.UTF_8)).use { reader ->
            return reader.readText()
        }
    }

    private fun parseMaxAge(header: String?): Int? {
        if (header.isNullOrBlank()) return null
        val parts = header.split(",")
        for (part in parts) {
            val trimmed = part.trim()
            if (trimmed.lowercase(Locale.US).startsWith("max-age=")) {
                val value = trimmed.substringAfter("=").toIntOrNull()
                if (value != null && value >= 0) {
                    return value
                }
            }
        }
        return null
    }

    private fun stripWeakEtag(etag: String?): String? {
        if (etag.isNullOrBlank()) return null
        var candidate = etag.trim()
        if (candidate.startsWith("W/", ignoreCase = true)) {
            candidate = candidate.substring(2).trim()
        }
        return candidate.trim('"')
    }

    data class Components(
        val slopeM: Double,
        val windM: Double,
        val tempM: Double,
        val altM: Double,
    )

    data class Result(
        val distanceEff: Double,
        val components: Components,
        val quality: Quality,
    )

    data class Options(
        val kS: Double = 1.0,
        val kHW: Double = 2.5,
        val warnThresholdRatio: Double = 0.05,
        val lowThresholdRatio: Double = 0.12,
        val config: Config? = null,
        val temperatureC: Double? = null,
        val altitudeAslM: Double? = null,
    )

    data class Config(
        val windModel: String = "percent_v1",
        val alphaHeadPerMph: Double = 0.01,
        val alphaTailPerMph: Double = 0.005,
        val slopeFactor: Double = 1.0,
        val windCapPctOfD: Double = 0.20,
        val taperStartMph: Double = 20.0,
        val sidewindDistanceAdjust: Boolean = false,
        val temperatureEnabled: Boolean = false,
        val betaTempPerC: Double = 0.0018,
        val altitudeEnabled: Boolean = false,
        val gammaAltPer100m: Double = 0.0065,
    )

    enum class Quality(val value: String) {
        GOOD("good"),
        WARN("warn"),
        LOW("low"),
    }

    fun computeSlopeAdjust(D: Double, deltaH: Double, kS: Double = 1.0): Double {
        if (!D.isFinite() || D <= 0.0) return 0.0
        if (!deltaH.isFinite()) return 0.0
        val clampedKs = kS.coerceIn(0.2, 3.0)
        return deltaH * clampedKs
    }

    fun computeWindAdjust(D: Double, wParallel: Double, kHW: Double = 2.5): Double {
        if (!D.isFinite() || D <= 0.0) return 0.0
        if (!wParallel.isFinite()) return 0.0
        val clamped = kHW.coerceIn(0.5, 6.0)
        return wParallel * clamped
    }

    fun computeWindAdjustPercentV1(
        D: Double,
        wParallel: Double,
        config: Config = Config(),
    ): Double {
        val distance = sanitizeDistance(D)
        if (distance <= 0.0) return 0.0
        if (!wParallel.isFinite() || wParallel == 0.0) return 0.0
        val windMph = abs(wParallel) * MPS_TO_MPH
        if (windMph == 0.0) return 0.0
        val taperStart = max(config.taperStartMph, 0.0)
        val isHeadwind = wParallel >= 0.0
        val alpha = max(if (isHeadwind) config.alphaHeadPerMph else config.alphaTailPerMph, 0.0)
        val below = minOf(windMph, taperStart) * alpha
        val above = max(windMph - taperStart, 0.0) * alpha * 0.8
        var pct = below + above
        if (!isHeadwind) {
            pct = -pct
        }
        val cap = max(config.windCapPctOfD, 0.0)
        pct = pct.coerceIn(-cap, cap)
        return distance * pct
    }

    fun computeTempAdjust(
        D: Double,
        temperatureC: Double,
        betaTemp: Double = 0.0018,
    ): Double {
        val distance = sanitizeDistance(D)
        if (distance <= 0.0 || !temperatureC.isFinite()) return 0.0
        val beta = if (betaTemp.isFinite()) betaTemp else 0.0018
        val delta = distance * beta * (20 - temperatureC)
        val cap = distance * 0.05
        return delta.coerceIn(-cap, cap)
    }

    fun computeAltitudeAdjust(
        D: Double,
        altitudeAslM: Double,
        gammaPer100m: Double = 0.0065,
    ): Double {
        val distance = sanitizeDistance(D)
        if (distance <= 0.0 || !altitudeAslM.isFinite()) return 0.0
        val gamma = if (gammaPer100m.isFinite()) gammaPer100m else 0.0065
        val delta = distance * gamma * (altitudeAslM / 100.0)
        val cap = distance * 0.15
        return delta.coerceIn(-cap, cap)
    }

    private fun computeQuality(distance: Double, deltaH: Double, wParallel: Double): Quality {
        if (distance <= 0.0) return Quality.LOW
        val hasSlope = deltaH.isFinite()
        val hasWind = wParallel.isFinite()
        if (!hasSlope && !hasWind) return Quality.LOW
        val windMph = if (hasWind) abs(wParallel) * MPS_TO_MPH else 0.0
        return if ((hasSlope && abs(deltaH) > 15.0) || windMph > 12.0) {
            Quality.WARN
        } else {
            Quality.GOOD
        }
    }

    fun computePlaysLike(
        D: Double,
        deltaH: Double,
        wParallel: Double,
        temperatureC: Double? = null,
        altitudeAslM: Double? = null,
        cfg: Config = Config(),
    ): Result {
        val distance = sanitizeDistance(D)
        val slope = computeSlopeAdjust(distance, deltaH, cfg.slopeFactor)
        val wind = if (cfg.windModel == "percent_v1") {
            computeWindAdjustPercentV1(distance, wParallel, cfg)
        } else {
            0.0
        }
        val temp = if (cfg.temperatureEnabled && temperatureC?.isFinite() == true) {
            computeTempAdjust(distance, temperatureC, cfg.betaTempPerC)
        } else {
            0.0
        }
        val alt = if (cfg.altitudeEnabled && altitudeAslM?.isFinite() == true) {
            computeAltitudeAdjust(distance, altitudeAslM, cfg.gammaAltPer100m)
        } else {
            0.0
        }
        val eff = distance + slope + wind + temp + alt
        val quality = computeQuality(distance, deltaH, wParallel)
        return Result(
            distanceEff = roundTo(eff, 1),
            components = Components(
                slopeM = roundTo(slope, 1),
                windM = roundTo(wind, 1),
                tempM = roundTo(temp, 1),
                altM = roundTo(alt, 1),
            ),
            quality = quality,
        )
    }

    fun compute(
        D: Double,
        deltaH: Double,
        wParallel: Double,
        opts: Options = Options(),
    ): Result {
        val baseConfig = opts.config ?: Config()
        val resolved = baseConfig.copy(slopeFactor = opts.kS)
        val result = computePlaysLike(
            D,
            deltaH,
            wParallel,
            opts.temperatureC,
            opts.altitudeAslM,
            resolved,
        )
        val distance = sanitizeDistance(D)
        val total =
            abs(result.components.slopeM) +
            abs(result.components.windM) +
            abs(result.components.tempM) +
            abs(result.components.altM)
        val ratio = if (distance > 0.0) total / distance else Double.POSITIVE_INFINITY
        return if (opts.warnThresholdRatio != 0.05 || opts.lowThresholdRatio != 0.12) {
            val quality = when {
                !ratio.isFinite() -> Quality.LOW
                ratio <= opts.warnThresholdRatio -> Quality.GOOD
                ratio <= opts.lowThresholdRatio -> Quality.WARN
                else -> Quality.LOW
            }
            result.copy(quality = quality)
        } else {
            result
        }
    }
}

File: shared/playslike/PlaysLikeService.ts
export type PlaysLikeQuality = "good" | "warn" | "low";

export interface PlaysLikeComponents {
  slopeM: number;
  windM: number;
}

export interface PlaysLikeResult {
  distanceEff: number;
  components: PlaysLikeComponents;
  quality: PlaysLikeQuality;
}

import literatureProfile from "../../tools/playslike/literature_v1.json";

export interface PlaysLikeOptions {
  kS?: number;
  kHW?: number;
  warnThresholdRatio?: number;
  lowThresholdRatio?: number;
  config?: Partial<PlaysLikeCfg>;
  clubClass?: string | null;
  playerType?: string | null;
}

interface CacheValueWithTtl {
  ttlSeconds: number;
}

type CacheEntry<T extends CacheValueWithTtl> = {
  value: T;
  etag?: string;
  expiresAt: number;
};

export interface ElevationProviderData extends CacheValueWithTtl {
  elevationM: number;
  etag?: string;
}

export interface WindProviderData extends CacheValueWithTtl {
  speedMps: number;
  dirFromDeg: number;
  wParallel: number | null;
  wPerp: number | null;
  etag?: string;
}

const elevationCache = new Map<string, CacheEntry<ElevationProviderData>>();
const windCache = new Map<string, CacheEntry<WindProviderData>>();

const STUB_ELEVATION: ElevationProviderData = {
  elevationM: 0,
  ttlSeconds: 0,
};

const STUB_WIND: WindProviderData = {
  speedMps: 0,
  dirFromDeg: 0,
  wParallel: 0,
  wPerp: 0,
  ttlSeconds: 0,
};

let providersBaseUrl: string | null = null;

const toKey = (lat: number, lon: number) =>
  `${lat.toFixed(5)},${lon.toFixed(5)}`;

const nowMs = () => Date.now();

const normalizeBaseUrl = (url: string | null): string | null => {
  if (!url) return null;
  return url.replace(/\/+$/, "");
};

const parseMaxAge = (header: string | null | undefined): number | undefined => {
  if (!header) return undefined;
  for (const token of header.split(",")) {
    const trimmed = token.trim().toLowerCase();
    if (trimmed.startsWith("max-age=")) {
      const value = Number.parseInt(trimmed.substring(8), 10);
      if (Number.isFinite(value) && value >= 0) {
        return value;
      }
    }
  }
  return undefined;
};

const stripWeakEtag = (etag: string | null | undefined): string | undefined => {
  if (!etag) return undefined;
  const trimmed = etag.trim();
  if (!trimmed) return undefined;
  const withoutWeak = trimmed.startsWith("W/") ? trimmed.substring(2).trim() : trimmed;
  return withoutWeak.replace(/^"|"$/g, "");
};

const updateEntryExpiry = <T extends CacheValueWithTtl>(
  entry: CacheEntry<T>,
  ttlSeconds: number | undefined,
) => {
  const ttl = Number.isFinite(ttlSeconds) ? Math.max(0, Math.floor(ttlSeconds ?? 0)) : 0;
  entry.value = { ...entry.value, ttlSeconds: ttl };
  entry.expiresAt = nowMs() + ttl * 1000;
  return ttl;
};

const buildUrl = (
  base: string,
  path: string,
  params: Record<string, string | number | null | undefined>,
) => {
  const query = Object.entries(params)
    .filter((entry): entry is [string, string | number] => {
      const value = entry[1];
      return value !== undefined && value !== null;
    })
    .map(([key, value]) => {
      const stringValue =
        typeof value === "number" ? value.toString() : String(value);
      return `${encodeURIComponent(key)}=${encodeURIComponent(stringValue)}`;
    })
    .join("&");
  return `${base}${path}${query ? `?${query}` : ""}`;
};

const refreshFrom304 = <T extends CacheValueWithTtl>(
  cache: Map<string, CacheEntry<T>>,
  key: string,
  entry: CacheEntry<T>,
  headers: Headers,
) => {
  const headerEtag = stripWeakEtag(headers.get("etag"));
  if (headerEtag) {
    entry.etag = headerEtag;
  }
  const ttl = parseMaxAge(headers.get("cache-control"));
  updateEntryExpiry(entry, ttl ?? entry.value.ttlSeconds);
  cache.set(key, entry);
  return { ...entry.value, etag: entry.etag };
};

export const setProvidersBaseUrl = (url: string | null) => {
  providersBaseUrl = normalizeBaseUrl(url);
};

export const getProvidersBaseUrl = () => providersBaseUrl;

const ensureBaseUrl = () => providersBaseUrl;

export const fetchElevation = async (
  lat: number,
  lon: number,
): Promise<ElevationProviderData> => {
  const base = ensureBaseUrl();
  if (!base) {
    return { ...STUB_ELEVATION };
  }
  const key = toKey(lat, lon);
  const cached = elevationCache.get(key);
  if (cached && cached.expiresAt > nowMs()) {
    return { ...cached.value, etag: cached.etag };
  }

  const headers: Record<string, string> = {};
  if (cached?.etag) {
    headers["If-None-Match"] = cached.etag;
  }

  const url = buildUrl(base, "/providers/elevation", { lat, lon });
  const response = await fetch(url, { headers });

  if (response.status === 304) {
    if (cached) {
      return refreshFrom304(elevationCache, key, cached, response.headers);
    }
    throw new Error("Received 304 without cached elevation entry");
  }

  if (!response.ok) {
    throw new Error(`Elevation provider error: ${response.status}`);
  }

  const payload = await response.json();
  const ttlCandidate =
    typeof payload.ttl_s === "number" ? payload.ttl_s : parseMaxAge(response.headers.get("cache-control"));
  const ttl = Math.max(0, Math.floor(ttlCandidate ?? 0));
  const entry: CacheEntry<ElevationProviderData> = {
    value: {
      elevationM: Number(payload.elevation_m ?? 0),
      ttlSeconds: ttl,
    },
    etag: payload.etag ?? stripWeakEtag(response.headers.get("etag")),
    expiresAt: nowMs() + ttl * 1000,
  };
  elevationCache.set(key, entry);
  return { ...entry.value, etag: entry.etag };
};

export const fetchWind = async (
  lat: number,
  lon: number,
  bearing?: number,
): Promise<WindProviderData> => {
  const base = ensureBaseUrl();
  if (!base) {
    return { ...STUB_WIND };
  }
  const key = `${toKey(lat, lon)}${bearing === undefined ? "" : `@${bearing.toFixed(2)}`}`;
  const cached = windCache.get(key);
  if (cached && cached.expiresAt > nowMs()) {
    return { ...cached.value, etag: cached.etag };
  }

  const headers: Record<string, string> = {};
  if (cached?.etag) {
    headers["If-None-Match"] = cached.etag;
  }

  const url = buildUrl(base, "/providers/wind", {
    lat,
    lon,
    bearing: bearing === undefined ? undefined : bearing,
  });
  const response = await fetch(url, { headers });

  if (response.status === 304) {
    if (cached) {
      return refreshFrom304(windCache, key, cached, response.headers);
    }
    throw new Error("Received 304 without cached wind entry");
  }

  if (!response.ok) {
    throw new Error(`Wind provider error: ${response.status}`);
  }

  const payload = await response.json();
  const ttlCandidate =
    typeof payload.ttl_s === "number" ? payload.ttl_s : parseMaxAge(response.headers.get("cache-control"));
  const ttl = Math.max(0, Math.floor(ttlCandidate ?? 0));
  const value: WindProviderData = {
    speedMps: Number(payload.speed_mps ?? 0),
    dirFromDeg: Number(payload.dir_from_deg ?? 0),
    wParallel:
      payload.w_parallel === null || payload.w_parallel === undefined
        ? null
        : Number(payload.w_parallel),
    wPerp:
      payload.w_perp === null || payload.w_perp === undefined ? null : Number(payload.w_perp),
    ttlSeconds: ttl,
  };
  const entry: CacheEntry<WindProviderData> = {
    value,
    etag: payload.etag ?? stripWeakEtag(response.headers.get("etag")),
    expiresAt: nowMs() + ttl * 1000,
  };
  windCache.set(key, entry);
  return { ...value, etag: entry.etag };
};

const clamp = (value: number, min: number, max: number) =>
  Math.min(Math.max(value, min), max);

const MPS_TO_MPH = 2.237;
const MPH_TO_MPS = 1 / MPS_TO_MPH;

export const mpsToMph = (value: number) => value * MPS_TO_MPH;
export const mphToMps = (value: number) => value * MPH_TO_MPS;

type LiteratureProfile = typeof literatureProfile;

type PlaysLikeScale = {
  scaleHead?: number;
  scaleTail?: number;
};

type PlaysLikeScaleMap = Record<string, PlaysLikeScale | undefined> | null | undefined;

export interface PlaysLikeCfg {
  windModel: "percent_v1";
  alphaHead_per_mph: number;
  alphaTail_per_mph: number;
  slopeFactor: number;
  windCap_pctOfD: number;
  taperStart_mph: number;
  sidewindDistanceAdjust: boolean;
  playsLikeProfile?: string;
  byClub?: PlaysLikeScaleMap;
  byPlayerType?: PlaysLikeScaleMap;
}

export const DEFAULT_PLAYSLIKE_CFG: PlaysLikeCfg = {
  windModel: "percent_v1",
  alphaHead_per_mph: 0.01,
  alphaTail_per_mph: 0.005,
  slopeFactor: 1.0,
  windCap_pctOfD: 0.2,
  taperStart_mph: 20,
  sidewindDistanceAdjust: false,
  playsLikeProfile: "literature_v1",
  byClub: (literatureProfile as LiteratureProfile).byClub,
  byPlayerType: (literatureProfile as LiteratureProfile).byPlayerType,
};

const roundTo = (value: number, decimals: number) => {
  if (!Number.isFinite(value)) return 0;
  const factor = 10 ** Math.max(0, decimals);
  return Math.round(value * factor) / factor;
};

const sanitizeDistance = (value: number) =>
  Number.isFinite(value) && value > 0 ? value : 0;

const resolveCfg = (cfg?: Partial<PlaysLikeCfg>): PlaysLikeCfg => {
  const merged: PlaysLikeCfg = { ...DEFAULT_PLAYSLIKE_CFG };
  if (!cfg) {
    return merged;
  }
  for (const [key, value] of Object.entries(cfg) as [keyof PlaysLikeCfg, unknown][]) {
    if (value !== undefined) {
      (merged as Record<string, unknown>)[key] = value;
    }
  }
  return merged;
};

type WindAlphas = Pick<PlaysLikeCfg, "alphaHead_per_mph" | "alphaTail_per_mph">;

const pickScale = (
  source: PlaysLikeScale | undefined,
): { head: number; tail: number } => ({
  head: source?.scaleHead ?? 1,
  tail: source?.scaleTail ?? 1,
});

export const applyLiteratureScaling = (
  clubClass: string | null | undefined,
  playerType: string | null | undefined,
  baseAlphas: WindAlphas,
  overrides?: { byClub?: PlaysLikeScaleMap; byPlayerType?: PlaysLikeScaleMap },
): WindAlphas => {
  const defaults: LiteratureProfile = literatureProfile;
  const clubMap = overrides?.byClub ?? defaults.byClub;
  const playerMap = overrides?.byPlayerType ?? defaults.byPlayerType;

  let scaleHead = 1;
  let scaleTail = 1;

  if (clubClass) {
    const scale = pickScale(clubMap?.[clubClass]);
    scaleHead *= scale.head;
    scaleTail *= scale.tail;
  }

  if (playerType) {
    const scale = pickScale(playerMap?.[playerType]);
    scaleHead *= scale.head;
    scaleTail *= scale.tail;
  }

  return {
    alphaHead_per_mph: baseAlphas.alphaHead_per_mph * scaleHead,
    alphaTail_per_mph: baseAlphas.alphaTail_per_mph * scaleTail,
  };
};

interface ComputePlaysLikeOptions {
  cfg?: Partial<PlaysLikeCfg>;
  clubClass?: string | null;
  playerType?: string | null;
}

export const computeSlopeAdjust = (
  D: number,
  deltaH: number,
  slopeFactor = DEFAULT_PLAYSLIKE_CFG.slopeFactor,
): number => {
  if (!Number.isFinite(D) || D <= 0 || !Number.isFinite(deltaH)) return 0;
  const gain = clamp(slopeFactor, 0.2, 3.0);
  return deltaH * gain;
};

type PercentV1Options = Pick<
  PlaysLikeCfg,
  "alphaHead_per_mph" | "alphaTail_per_mph" | "windCap_pctOfD" | "taperStart_mph"
>;

export const computeWindAdjustPercentV1 = (
  D: number,
  wParallel_mps: number,
  opts?: Partial<PercentV1Options>,
): number => {
  const distance = sanitizeDistance(D);
  if (distance <= 0 || !Number.isFinite(wParallel_mps) || wParallel_mps === 0) {
    return 0;
  }
  const alphaHead = Math.max(
    opts?.alphaHead_per_mph ?? DEFAULT_PLAYSLIKE_CFG.alphaHead_per_mph,
    0,
  );
  const alphaTail = Math.max(
    opts?.alphaTail_per_mph ?? DEFAULT_PLAYSLIKE_CFG.alphaTail_per_mph,
    0,
  );
  const windCap = Math.max(
    opts?.windCap_pctOfD ?? DEFAULT_PLAYSLIKE_CFG.windCap_pctOfD,
    0,
  );
  const taperStart = Math.max(
    opts?.taperStart_mph ?? DEFAULT_PLAYSLIKE_CFG.taperStart_mph,
    0,
  );
  const windMph = Math.abs(wParallel_mps) * MPS_TO_MPH;
  if (windMph === 0) return 0;
  const isHeadwind = wParallel_mps >= 0;
  const alpha = isHeadwind ? alphaHead : alphaTail;
  const below = Math.min(windMph, taperStart) * alpha;
  const above = Math.max(windMph - taperStart, 0) * alpha * 0.8;
  let pct = below + above;
  if (!isHeadwind) {
    pct = -pct;
  }
  const cappedPct = clamp(pct, -windCap, windCap);
  return distance * cappedPct;
};

const computeQualityFromInputs = (
  distance: number,
  deltaH: number,
  wParallel_mps: number,
): PlaysLikeQuality => {
  if (distance <= 0) return "low";
  const hasSlope = Number.isFinite(deltaH);
  const hasWind = Number.isFinite(wParallel_mps);
  if (!hasSlope && !hasWind) return "low";
  const windMph = hasWind ? Math.abs(wParallel_mps) * MPS_TO_MPH : 0;
  if ((hasSlope && Math.abs(deltaH) > 15) || windMph > 12) {
    return "warn";
  }
  return "good";
};

const toComputeOptions = (
  value?: Partial<PlaysLikeCfg> | ComputePlaysLikeOptions,
): ComputePlaysLikeOptions => {
  if (!value) return {};
  if (
    typeof value === "object" &&
    ("cfg" in value || "clubClass" in value || "playerType" in value)
  ) {
    return value as ComputePlaysLikeOptions;
  }
  return { cfg: value as Partial<PlaysLikeCfg> };
};

export const computePlaysLike = (
  D: number,
  deltaH: number,
  wParallel_mps: number,
  options?: Partial<PlaysLikeCfg> | ComputePlaysLikeOptions,
): PlaysLikeResult => {
  const distance = sanitizeDistance(D);
  const { cfg, clubClass, playerType } = toComputeOptions(options);
  const resolved = resolveCfg(cfg);
  const slope = computeSlopeAdjust(distance, deltaH, resolved.slopeFactor);
  const baseAlphas: WindAlphas = {
    alphaHead_per_mph: resolved.alphaHead_per_mph,
    alphaTail_per_mph: resolved.alphaTail_per_mph,
  };
  const scaledAlphas =
    resolved.playsLikeProfile === "literature_v1"
      ? applyLiteratureScaling(clubClass, playerType, baseAlphas, {
          byClub: resolved.byClub,
          byPlayerType: resolved.byPlayerType,
        })
      : baseAlphas;
  const wind =
    resolved.windModel === "percent_v1"
      ? computeWindAdjustPercentV1(distance, wParallel_mps, {
          ...resolved,
          ...scaledAlphas,
        })
      : 0;
  const eff = distance + slope + wind;
  const quality = computeQualityFromInputs(distance, deltaH, wParallel_mps);
  return {
    distanceEff: roundTo(eff, 1),
    components: {
      slopeM: roundTo(slope, 1),
      windM: roundTo(wind, 1),
    },
    quality,
  };
};

export const computeWindAdjust = (D: number, wParallel: number, kHW = 2.5): number => {
  if (!Number.isFinite(D) || D <= 0 || !Number.isFinite(wParallel)) return 0;
  const gain = clamp(kHW, 0.5, 6.0);
  return wParallel * gain;
};

export const compute = (
  D: number,
  deltaH: number,
  wParallel: number,
  opts: PlaysLikeOptions = {},
): PlaysLikeResult => {
  const overrides: Partial<PlaysLikeCfg> = { ...(opts.config ?? {}) };
  if (opts.kS !== undefined) {
    overrides.slopeFactor = opts.kS;
  }
  const result = computePlaysLike(D, deltaH, wParallel, {
    cfg: overrides,
    clubClass: opts.clubClass,
    playerType: opts.playerType,
  });
  const distance = sanitizeDistance(D);
  const total = Math.abs(result.components.slopeM) + Math.abs(result.components.windM);
  const ratio = distance > 0 ? total / distance : Number.POSITIVE_INFINITY;
  const warnThreshold = opts.warnThresholdRatio;
  const lowThreshold = opts.lowThresholdRatio;
  if (warnThreshold !== undefined || lowThreshold !== undefined) {
    const warn = warnThreshold ?? 0.05;
    const low = lowThreshold ?? 0.12;
    let quality: PlaysLikeQuality;
    if (!Number.isFinite(ratio)) {
      quality = "low";
    } else if (ratio <= warn) {
      quality = "good";
    } else if (ratio <= low) {
      quality = "warn";
    } else {
      quality = "low";
    }
    return {
      ...result,
      quality,
    };
  }
  return result;
};

export const mergePlaysLikeCfg = (cfg?: Partial<PlaysLikeCfg>): PlaysLikeCfg =>
  resolveCfg(cfg);
